src/samples/http/server.c

```diff
@@ -0,0 +1,179 @@
+ *
+ * Copyright (C) 2015 Intel Corporation. All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ *
+ *   * Redistributions of source code must retain the above copyright
+ *     notice, this list of conditions and the following disclaimer.
+ *   * Redistributions in binary form must reproduce the above copyright
+ *     notice, this list of conditions and the following disclaimer in
+ *     the documentation and/or other materials provided with the
+ *     distribution.
+ *   * Neither the name of Intel Corporation nor the names of its
+ *     contributors may be used to endorse or promote products derived
+ *     from this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include <fcntl.h>
+#include <stdio.h>
+#include <string.h>
+#include <sys/stat.h>
+#include <sys/types.h>
+#include <unistd.h>
+
+#include "sol-mainloop.h"
+#include "sol-http.h"
+#include "sol-http-server.h"
+
+static struct sol_http_server *server;
+static struct sol_fd *stdin_watch;
+static char *value;
+
+static bool
+on_stdin(void *data, int fd, uint32_t flags)
+{
+    ssize_t r;
+    char buffer[512];
+
+    if (flags & SOL_FD_FLAGS_ERR) {
+        fprintf(stderr, "ERROR: Something wrong happened with file descriptor: %d\n", fd);
+        goto err;
+    }
+
+    if (flags & SOL_FD_FLAGS_IN) {
+        r = read(fd, buffer, sizeof(buffer) - 1);
+        if (r  == -1) {
+            if ((errno == EAGAIN) || (errno == EINTR))
+                return true;
+            fprintf(stderr,
+                "ERROR: Failed to read file descriptor: %d\n", fd);
+            goto err;
+        }
+
+        buffer[r] = '\0';
+        free(value);
+        value = strdup(buffer);
+    }
+
+    return true;
+
+err:
+    sol_quit_with_code(EXIT_FAILURE);
+    return false;
+}
+
+static int
+request_cb(void *data, struct sol_http_request *request)
+{
+    int r;
+    struct sol_http_response response = {
+        SOL_SET_API_VERSION(.api_version = SOL_HTTP_RESPONSE_API_VERSION, )
+        .content = SOL_BUFFER_INIT_EMPTY,
```

Comment:
since we're using pointers we're not going to change and they will live during this function (as `value` may be modified later, I'm assuming that), then we can use different macro to initialize the buffer to a memory-not-owned value and give the data + size, be the constant string or the `value` (which should be a buffer anyway).